### DISCUSSION TITLE
How to use the Callback function for the Hammerspoon Virtual MIDI object ( analogue with hs.midi.new)

### DISCUSSION THREAD

**Discussion by Vladistone** [Hammerspoon/hammerspoon#3432] [Q&A]
HS console review: (but is doesn't working... where I make error with virtualSource?
```
2023-03-17 19:04:29: Welcome to the Hammerspoon Console!
You can run any Lua code in here.

2023-03-17 19:04:29: -- Lazy extension loading enabled
2023-03-17 19:04:29: -- Loading ~/.hammerspoon/init.lua
2023-03-17 19:04:29: -- Done.
----
2023-03-17 19:04:36: -- Loading extension: inspect
2023-03-17 19:04:36: -- Loading extension: midi
2023-03-17 19:04:36: { "network", "Bluetooth", "ipMIDI" }
2023-03-17 19:04:36: { "Port 1", "Port 2", "Port 3", "Port 4", "Port 5", "Port 6" }
2023-03-17 19:04:36: -- Loading extension: window
2023-03-17 19:04:36: -- Loading extensions: window.filter
2023-03-17 19:04:36: -- Loading extension: osascript
```
script:

```
-- print available MIDI device names to the Hammerspoon Console
print(hs.inspect(hs.midi.devices()))
virtualSources = hs.midi.virtualSources()
print(hs.inspect(virtualSources))

-- Create a new hs.midi object for the desired MIDI device by passing in the device name
local Control = hs.midi.new("ipMIDI")
local virtControl = hs.midi.newVirtualSource("Port 3")
local windowFilter = hs.window.filter
local runScript = hs.osascript


-- AppleScript to click plugin Compare button
local clickCompareButton = [[
  tell application "System Events"
	  tell process "Logic Pro"
        click checkbox "Compare" of window 1
	  end tell
  end tell
]]

-- Callback function for the Hammerspoon MIDI object created above (hs.midi.new)
-- Parameters: object, deviceName, commandType, description, metadata
-- Here we listen for incoming Control Change message that will trigger the AppleScript when received
local handleMidi = function(...)
  local commandType = select(3, ...)
  local metaData = select(5, ...)
  if commandType == 'controlChange' then
      if metaData.controlNumber == 39 and metaData.controlValue == 127 then
        runScript.applescript(clickCompareButton)
      end
  end
end


-- Create Window Filter for Logic Pro window
local logicWindowFilter = windowFilter.new('Logic Pro')

-- Subscribe to events when Logic Pro window is in focus (or not)
logicWindowFilter:subscribe(windowFilter.windowFocused, function()
  -- Sets the callback for the Hammerspoon MIDI object when Logic Pro is in focus
  Control:callback(handleMidi)
  print('Logic in focus...')
end):subscribe(windowFilter.windowUnfocused, function()
  -- Removes the callback when Logic Pro loses focus
  Control:callback(nil)
  print('Logic not in focus...')
end)
```

**Comment by cmsj** [comment]
@latenitefilms may be able to help here?

**Reply by latenitefilms** [reply]
It's a little hard to follow, but it doesn't look like you actually have a callback on your `virtControl = hs.midi.newVirtualSource("Port 3")`? You only seem to have a callback on `Control = hs.midi.new("ipMIDI")`?

It doesn't seem like you're currently doing anything with `virtControl`?

**Comment by Vladistone** [comment]
pardon for my lua skills (12 hours stady old )))
I didn't understand virtual encoding operations
https://www.hammerspoon.org/docs/hs.midi.html#commandTypes
when I checked print parameters by code:
```-- Checking the open midi devices and virtual port name of ipMIDI:

print("MIDI devices: " .. hs.inspect(hs.midi.devices()))
midiDevice = hs.midi.virtualSources()
print("Virtual Sources   1:" .. hs.inspect(midiDevice))

-- Creating the virtual var.:

midiDevice = hs.midi.new(hs.midi.virtualSources()[1])
print("Virtual Sources new: " .. hs.inspect(midiDevice))

-- Callback function for watching of receive MIDI events from ipMIDI Controller:

midiDevice:callback(function(object, deviceName, commandType, description, metadata)
           print("object: " .. tostring(object))
           print("deviceName: " .. deviceName)
           print("commandType: " .. commandType)
           print("description: " .. description)
           print("metadata: " .. hs.inspect(metadata))
           end)
```
 I recive only "nil&error":

```
2023-03-17 23:03:13: Welcome to the Hammerspoon Console!
You can run any Lua code in here.

2023-03-17 23:03:13: -- Lazy extension loading enabled
2023-03-17 23:03:13: -- Loading ~/.hammerspoon/init.lua
2023-03-17 23:03:13: -- Loading extension: inspect
2023-03-17 23:03:13: -- Loading extension: midi
2023-03-17 23:03:13: MIDI devices: { "network", "Bluetooth", "ipMIDI" }
2023-03-17 23:03:13: Virtual Sources   1:{ "Port 1", "Port 2", "Port 3", "Port 4", "Port 5", "Port 6" }
2023-03-17 23:03:13: Virtual Sources new: nil
2023-03-17 23:03:13: *** ERROR: /Users/vladistone/.hammerspoon/init.lua:6: attempt to index a nil value (global 'midiDevice')
stack traceback:
	/Users/vladistone/.hammerspoon/init.lua:6: in main chunk
	[C]: in function 'xpcall'
	...app/Contents/Resources/extensions/hs/_coresetup/init.lua:722: in function 'hs._coresetup.setup'
	(...tail calls...)
```

where is my fault or error?

**Reply by latenitefilms** [reply]
Both `hs.midi.new(deviceName)` and `hs.midi.newVirtualSource(virtualSource)` are Constructors that behave the same way. They both accept a single string parameter and return a `hs.midi` object.

For example, `hs.midi.new(hs.midi.devices()[1])` will return a `hs.midi` object, as will `hs.midi.newVirtualSource(hs.midi.virtualSources()[1])`.

To create a new Virtual Source MIDI object for "Port 1" you could do something like:

```
myVirtualPortOne = hs.midi.newVirtualSource("Port 1")
print(string.format("myVirtualPortOne: %s", myVirtualPortOne))
myVirtualPortOne:callback(function(object, deviceName, commandType, description, metadata)
           print("object: " .. tostring(object))
           print("deviceName: " .. deviceName)
           print("commandType: " .. commandType)
           print("description: " .. description)
           print("metadata: " .. hs.inspect(metadata))
end)
```

**Comment by Vladistone** [comment]
You helped me tremendously with the learning curve.
but in the documentation example, I found a different version of the code ... or is it a typo?
![virt MIDI](https://user-images.githubusercontent.com/91404765/225940115-3894c4da-8935-4bd8-bc47-9e2149722daa.png)

**Reply by Vladistone** [reply]
> You helped me tremendously with the learning curve. but in the documentation example, I found a different version of the code ... or is it a typo? ![virt MIDI](https://user-images.githubusercontent.com/91404765/225940115-3894c4da-8935-4bd8-bc47-9e2149722daa.png)
 any comments?

**Reply by latenitefilms** [reply]
Apologies - that's a typo. Have corrected in #3435.
